[TIMOB-6305] iOS: Add Stepper UI element

AngelkPetkov opened the pull request:
[JIRA](https://jira.appcelerator.org/browse/TIMOB-6305) 
Adding Support for UIStepper

AngelkPetkov committed:
[TIMOB-6305] iOS: Add Stepper UI element

Add supprot for UIStepper

AngelkPetkov committed:
[TIMOB-6305] iOS: Moved to iOS tag

AngelkPetkov committed:
[TIMOB-6405] iOS: Adding Docs

AngelkPetkov committed:
[TIMOB-6305] iOS: Fixed yml error and added events

hansemannn commented on apidoc/Titanium/UI/iOS/Stepper.yml:
```diff
@@ -0,0 +1,180 @@
+summary: |
+    A widget used to increment and decrement a value
+description: |
+    <table id="IOS Stepper">
```
An `id` is not allowed to have spaces. And why is a HTML id even required here?

hansemannn commented on apidoc/Titanium/UI/iOS/Stepper.yml:
```diff
@@ -0,0 +1,180 @@
+description: |
+    <table id="IOS Stepper">
+      <tr>
+        <td><img src="images/stepper/stepper.png" height="50" /></td>
```
You calling the directory "stepper", but the below folder is called "Stepper"

hansemannn commented on apidoc/Titanium/UI/iOS/Stepper.yml:
```diff
@@ -0,0 +1,180 @@
+    <table id="IOS Stepper">
+      <tr>
+        <td><img src="images/stepper/stepper.png" height="50" /></td>
+        <td><img src="images/stepper/stepper_customImages" height="51" /></td>
```
Missing ".png" suffix. Did you test the generated docs before?

hansemannn commented on apidoc/Titanium/UI/iOS/Stepper.yml:
```diff
@@ -0,0 +1,180 @@
+
+    A Stepper displays its selected state while it is being pressed.
+    You can specify background images for each state on iOS.
+    Use the <Titanium.UI.createstepper> method or a **`<stepper>`** Alloy element to create a stepper 
```
It's `createStepper`, not `createstepper` (It will cause JS runtime errors). Also `<Stepper>` instead of `<stepper>`. And did you implement the Alloy-part of this PR? I can't see any Alloy-related code+docs.

hansemannn commented on apidoc/Titanium/UI/iOS/Stepper.yml:
```diff
@@ -0,0 +1,180 @@
+    A Stepper displays its selected state while it is being pressed.
+    You can specify background images for each state on iOS.
+    Use the <Titanium.UI.createstepper> method or a **`<stepper>`** Alloy element to create a stepper 
+    (see [Examples](#!/api/Titanium.UI.iOS.stepper-examples)) .
```
Link won't work, because there is no `Titanium.UI.iOS.stepper`.

hansemannn commented on apidoc/Titanium/UI/iOS/Stepper.yml:
```diff
@@ -0,0 +1,180 @@
+    * Similarly, for an assigned background gradient to show through, the `backgroundColor` may need to be
+    set to 'transparent'.
+
+    *Note that the stepper cannot be re-sized.
```
Missing space between the work and the `*`. It will cause the list to break in generation

hansemannn commented on apidoc/Titanium/UI/iOS/Stepper.yml:
```diff
@@ -0,0 +1,180 @@
+since: "5.4.0"
+excludes:
+    events: [dblclick, doubletap, focus, keypressed, longclick, longpress, pinch, postlayout,
+             singletap, swipe, touchcancel, touchend, touchmove, touchstart, twofingertap]
```
Did you verify, that all these events don't work with the stepper? I would expect that at least `doubleclick` and the `touch*` events work like with other UIControls.

hansemannn commented on apidoc/Titanium/UI/iOS/Stepper.yml:
```diff
@@ -0,0 +1,180 @@
+    summary: The current value of the stepper.
+    type: Number
+
+  - name: continous
```
It's called `continuous`

hansemannn commented on apidoc/Titanium/UI/iOS/Stepper.yml:
```diff
@@ -0,0 +1,180 @@
+
+  - name: wraps
+    summary: |
+        If YES, incrementing beyond maximumValue sets value to minimumValue; likewise, decrementing below
```
Code-highlight `maximumValue ` and `minimumValue `. Use A point instead of a semicolon.

hansemannn commented on apidoc/Titanium/UI/iOS/Stepper.yml:
```diff
@@ -0,0 +1,180 @@
+    summary: The miniumum value the stepper will be set to, the value must be smaller than the maximumValue.
+    type: Number
+
+  - name: minimumValue
```
Copy-pasted. It's `maximumValue`

hansemannn commented on apidoc/Titanium/UI/iOS/Stepper.yml:
```diff
@@ -0,0 +1,180 @@
+    summary: The maximum value the stepper will be set to, the value must be greater than the miniumumValue.
+    type: Number
+
+  - name: minimumValue
```
Not sure what this should be, but it's not `minimumValue` again.

hansemannn commented on apidoc/Titanium/UI/iOS/Stepper.yml:
```diff
@@ -0,0 +1,180 @@
+    type: Number
+
+  - name: minimumValue
+    summary: The maximum value the stepper will be set to, the value must be greater than the miniumumValue.
```
What happens if the value is equal to the `minimumValue`?

hansemannn commented on apidoc/Titanium/UI/iOS/Stepper.yml:
```diff
@@ -0,0 +1,180 @@
+  - name: tintColor
+    summary: Sets the color for the widget, any backgroundImages added will be set to the same color.
+    description: |
+        Default color is blue
```
Missing point. Also think of removing this sentence, because the default tintColor depends on the global tint-configuration set. See https://developer.apple.com/library/ios/documentation/UserExperience/Conceptual/TransitionGuide/AppearanceCustomization.html

hansemannn commented on apidoc/Titanium/UI/iOS/Stepper.yml:
```diff
@@ -0,0 +1,180 @@
+        is unset, the image will be darkened to indicate being selected.
+    type: String
+
+  - name: backgroundDisabledImage
```
Did you test this? Cannot see a `enabled` property that would cause this state. Please add tests for all states to the FT-example.

hansemannn commented on apidoc/Titanium/UI/iOS/Stepper.yml:
```diff
@@ -0,0 +1,180 @@
+        Also sets the background image for the other stepper states (normal, focused, selected), 
+        unless they've been specified explicitly, except for iOS, which requires a backgroundImage.
+    type: String
+    platforms: [android, iphone, ipad, mobileweb]
```
Why are these platforms here? Isn't it iOS-only?

hansemannn commented on apidoc/Titanium/UI/iOS/Stepper.yml:
```diff
@@ -0,0 +1,180 @@
+        Background image for the stepper in its selected state, specified as a local file 
+        path or URL.
+    type: String
+    platforms: [iphone, ipad]
```
Similar as above: As the API is iOS-only, we don't need to explicitly declare the platform support

hansemannn commented on apidoc/Titanium/UI/iOS/Stepper.yml:
```diff
@@ -0,0 +1,180 @@
+        Also sets the background image for the other stepper states (disabled, focused, selected),
+        unless they've been specified explicitly, except for iOS, which requires a backgroundImage.
+    type: String
+    platforms: [iphone, ipad]
```
Same as above. Please remove the platforms in all other below occurrences as well.

hansemannn commented on apidoc/Titanium/UI/iOS/Stepper.yml:
```diff
@@ -0,0 +1,180 @@
+        unless they've been specified explicitly. On iOS, if backgroundDisabledImage is unset, the
+        image will appear faded to indicate being disabled. Also on iOS, if backgroundSelectedImage
+        is unset, the image will be darkened to indicate being selected.
+    type: String
```
What happens if I use a TiBlob/TiFile?

hansemannn commented on apidoc/Titanium/UI/iOS/Stepper.yml:
```diff
@@ -0,0 +1,180 @@
+
+  - name: incrementDisabledImage
+    summary: |
+        Background image for the stepper increment button in its dissabled state, specified as a local
```
Typo: "dissabled" -> "disabled"

hansemannn commented on apidoc/Titanium/UI/iOS/Stepper.yml:
```diff
@@ -0,0 +1,180 @@
+
+  - name: decrementDisabledImage
+    summary: |
+        Background image for the stepper decrement button in its dissabled state, specified as a local
```
Typo: "dissabled" -> "disabled"

hansemannn commented on apidoc/Titanium/UI/iOS/Stepper.yml:
```diff
@@ -0,0 +1,180 @@
+  - name: change
+    summary: Fired every time the stepper value changes.
+    properties:
+      - name: Value
```
Lowercase: "Value" -> value

hansemannn commented on apidoc/Titanium/UI/iOS/Stepper.yml:
```diff
@@ -0,0 +1,180 @@
+events:
+  - name: change
+    summary: Fired every time the stepper value changes.
+    properties:
```
Any other properties in that event that would make sense?

hansemannn commented on apidoc/Titanium/UI/iOS/Stepper.yml:
```diff
@@ -0,0 +1,180 @@
+    example: |
+
+            var win = Ti.UI.createWindow({
+            backgroundColor: '#fff'
```
Indentation. Please also fix the other lines below

hansemannn commented on iphone/Classes/TiUIiOSProxy.m:
```diff
@@ -715,6 +719,13 @@ -(id)createBlurView:(id)args
+#ifdef USE_TI_UIIOSSTEPPER
+-(id)createStepper:(id)args
+{
+    return [[[TiUIiOSStepperProxy alloc] _initWithPageContext:[self executionContext] args:args] autorelease];
```
Why not `[self pageContext]`? I remember we should use that for newer API's, but they should work the same.

AngelkPetkov replied:
I was just making it the same one as all the classes above including the blurView, will change it back to [self pageContext].

hansemannn commented on iphone/Classes/TiUIiOSStepper.h:
```diff
@@ -0,0 +1,18 @@
+@interface TiUIiOSStepper : TiUIView {
+@private
+    UIStepper *stepper;
+    UIImage *backgroundImageCache;
```
So here are my thoughts on this: Since we subclass the `TiUIView` in every UI element, I think we could do all the backgroundImage stuff on that view instead of adding another view that slows down the UI creation. Doing that, we wouldn't even need all the custom setter and logic to add the image, because the `TiUIView` already does all that by default. 

Please test this and verify it's necessary to do all this.

AngelkPetkov replied:
if i don't store the image inside a variable before setting it , it doesn't seem to want to get displayed.  Im following the same logic we do inside the button class.

hansemannn commented on iphone/Classes/TiUIiOSStepper.m:
```diff
@@ -0,0 +1,204 @@
+
+-(UIStepper*)stepper
+{
+    if (stepper ==nil)
```
Missing space in `==nil`.

hansemannn commented on iphone/Classes/TiUIiOSStepper.m:
```diff
@@ -0,0 +1,204 @@
+{
+    if (stepper ==nil)
+    {
+        stepper = [[[UIStepper alloc]init] retain];
```
Missing space in `alloc]init]`. Also, why do we need to manaully retain the UI element? Don't know of a case where it would be nil here.

hansemannn commented on iphone/Classes/TiUIiOSStepper.m:
```diff
@@ -0,0 +1,204 @@
+    if (stepper ==nil)
+    {
+        stepper = [[[UIStepper alloc]init] retain];
+        [stepper setMaximumValue:100];
```
Use the proxy value from `maximumValue` here instead of static values.

AngelkPetkov replied:
Ive added this so if the user doesn't set a value this will be the default one. If a user sets a value it will be set and the value set here will be overwritten.

hansemannn commented on iphone/Classes/TiUIiOSStepper.m:
```diff
@@ -0,0 +1,204 @@
+    {
+        stepper = [[[UIStepper alloc]init] retain];
+        [stepper setMaximumValue:100];
+        [stepper setMinimumValue:0];
```
Use the proxy value from `minimumValue ` here instead of static values.

hansemannn commented on iphone/Classes/TiUIiOSStepper.m:
```diff
@@ -0,0 +1,204 @@
+}
+-(void)setContinuous_:(id)value
+{
+    [[self stepper] setContinuous:[TiUtils boolValue:value]];
```
Validate the user input before: `ENSURE_TYPE(value, NSNumber);`

hansemannn commented on iphone/Classes/TiUIiOSStepper.m:
```diff
@@ -0,0 +1,204 @@
+
+-(void)setAutorepeat_:(id)value
+{
+    [[self stepper] setAutorepeat:[TiUtils boolValue:value]];
```
Same as above, missing input validation.

hansemannn commented on iphone/Classes/TiUIiOSStepper.m:
```diff
@@ -0,0 +1,204 @@
+
+-(void)setWraps_:(id)value
+{
+   [[self stepper] setWraps:[TiUtils boolValue:value]];
```
Same as above, missing input validation.

hansemannn commented on iphone/Classes/TiUIiOSStepper.m:
```diff
@@ -0,0 +1,204 @@
+
+-(void)setMaximumValue_:(id)value
+{
+    ENSURE_TYPE(value, NSNumber)
```
I like to do a closing semicolon after macros to unify the style with other methods in Obj-C, but that's up to you :blush:

hansemannn commented on iphone/Classes/TiUIiOSStepper.m:
```diff
@@ -0,0 +1,204 @@
+-(void)setStepValue_:(id)value
+{
+    ENSURE_TYPE(value, NSNumber)
+    NSNumber *stepValue = (NSNumber*)value;
```
Use `NSNumber *stepValue = [TiUtils doubleValue: value]` instead.

AngelkPetkov replied:
You cant initialize an NSNumber with a simple C type such as double , which is why I'm casting it :P. Thats what i actually tried first but the compiler didn't like it.

hansemannn replied:
Pretty sure you can, since it calls the double-initializer. We do that in many places.

AngelkPetkov replied:
No you definitely cant do it by doing NSNumber *stepValue = [TiUtils doubleValue: value]. I tried it before and just tried it again you get an error.

hansemannn replied:
I missed the constructor I talked about above:
```
NSNumber *stepValue = [NSNumber numberWithDouble:[TiUtils doubleValue:value]];
```

hansemannn replied:
So you can call `[[self stepper] setStepValue:[NSNumber numberWithDouble:[TiUtils doubleValue:value]]];` directly

hansemannn commented on iphone/Classes/TiUIiOSStepper.m:
```diff
@@ -0,0 +1,204 @@
+    if (stepValue > 0) {
+        [[self stepper] setStepValue:[TiUtils doubleValue:value]];
+    } else {
+        NSLog(@"[WARN]: The stepValue must be bigger than 0");
```
Remove the `:` and don't forget the point at the end.

hansemannn commented on iphone/Classes/TiUIiOSStepper.m:
```diff
@@ -0,0 +1,204 @@
+}
+-(void)setTintColor_:(id)value
+{
+    [[self stepper] setTintColor:[[TiUtils colorValue:value] color]];
```
Missing input validation for `TiColor` values.

hansemannn commented on iphone/Classes/TiUIiOSStepper.m:
```diff
@@ -0,0 +1,204 @@
+}
+
+#pragma mark backgroundImage
+-(void)setBackgroundImage_:(id)value
```
Missing blank-line after the `#pragma`

hansemannn replied:
As discussed above, think of doing all the image-stuff using the superclass instead of re-implementing it.

hansemannn commented on iphone/Classes/TiUIiOSStepper.m:
```diff
@@ -0,0 +1,204 @@
+}
+
+#pragma mark Decrement backgroundImage
+-(void)setDecrementImage_:(id)value
```
Missing blank-line.

hansemannn commented on iphone/Classes/TiUIiOSStepper.m:
```diff
@@ -0,0 +1,204 @@
+    [[self stepper] setDecrementImage:[self loadImage:value] forState:UIControlStateDisabled];
+}
+
+#pragma mark Increment backgroundImage
```
Missing blank-line.

hansemannn commented on iphone/Classes/TiUIiOSStepper.m:
```diff
@@ -0,0 +1,204 @@
+#pragma mark backgroundImage
+-(void)setBackgroundImage_:(id)value
+{
+    backgroundImageCache = [self imageWithImage:[self loadImage:value] fullScale:YES];
```
Validate the input before (if you decide to use this solution)

hansemannn commented on iphone/Classes/TiUIiOSStepper.m:
```diff
@@ -0,0 +1,204 @@
+- (IBAction)stepperChanged:(id)sender
+{
+    NSNumber *newValue = [NSNumber numberWithDouble:[[self stepper] value]];
+    [self.proxy replaceValue:newValue forKey:@"value" notification:YES];
```
Make sure you also call this method on every other proxy-setter like `mininumValue` etc. Otherwise, the properties won't be accessible after creating the proxy e.g. `stepper.mininumValue` etc.

AngelkPetkov replied:
This is called every time you decrement or increment , I'm not sure what you mean ? Changing the maximum value or minimum shouldn't effect the current value unless the maximum value is smaller than the current value or the minimum is larger. In which case it will change internally and then this will be called. Or if not soon as you increment/decrement it will be updated.

hansemannn commented on iphone/Classes/TiUIiOSStepper.m:
```diff
@@ -0,0 +1,204 @@
+    }
+}
+
+- (IBAction)stepperTouch:(id)sender forEvent:(UIEvent *)event
```
Isn't the click event handled using the super-class (`TiUIView`)?

hansemannn commented on iphone/Classes/TiUIiOSStepper.m:
```diff
@@ -0,0 +1,204 @@
+    UITouch *touch = [[event allTouches] anyObject];
+    if([self.proxy _hasListeners:@"click"]) {
+        [self.proxy fireEvent:@"click" withObject:[NSDictionary
+                                                   dictionaryWithDictionary:[TiUtils touchPropertiesToDictionary:touch
```
The format looks messed-up

hansemannn commented on iphone/Classes/TiUIiOSStepper.m:
```diff
@@ -0,0 +1,204 @@
+    //UIGraphicsBeginImageContext(newSize);
+    // In next line, pass 0.0 to use the current device's pixel scaling factor (and thus account for Retina resolution).
+    // Pass 1.0 to force exact pixel size.
+    [self stepper];
```
We talked about this during the fixes for ListView: Don't call the getter before you use it. Or: Think about a case where the stepper could be nil in this case and initialize it before this place.

AngelkPetkov replied:
Thats there uncase this somehow gets called while the stepper is nil , which did happen and caused crash, this way if nil it will be initialized.

hansemannn commented on iphone/Classes/TiUIiOSStepper.m:
```diff
@@ -0,0 +1,204 @@
+        UIImage *image = UIGraphicsGetImageFromCurrentImageContext();
+        UIGraphicsEndImageContext();
+        [theView release];
+        return [image retain];
```
Is this logic from somewhere else? Where do you release the `image`?

hansemannn commented on iphone/Classes/TiUIiOSStepperProxy.m:
```diff
@@ -0,0 +1,49 @@
+-(void)_initWithProperties:(NSDictionary *)properties
+{
+    [super _initWithProperties:properties];
+}
```
Not necessary, if you don't do any custom logic inside the proxy constructor.

hansemannn commented on iphone/Classes/TiUIiOSStepperProxy.m:
```diff
@@ -0,0 +1,49 @@
+
+-(UIView *)parentViewForChild:(TiViewProxy *)child
+{
+    return [[(TiUIiOSStepper*)[self view] stepper] superview];
```
Necessary? What does it do?

hansemannn commented on iphone/Classes/TiUIiOSStepperProxy.m:
```diff
@@ -0,0 +1,49 @@
+-(NSNumber*)value
+{
+    NSNumber *value = [NSNumber numberWithDouble:[[(TiUIiOSStepper*)[self view] stepper] value]];
+    return value;
```
Create a getter for your `TiUIiOSStepper` class and then call it with:
```objc
return NUMDOUBLE([[[self stepper] stepper] value]);
```

hansemannn commented:
Left some comments, will do the FT after the comments are addressed.

hansemannn commented on apidoc/Titanium/UI/iOS/Stepper.yml:
```diff
@@ -5,35 +5,40 @@ summary: |
 since: "5.4.0"
 excludes:
-    events: [dblclick, doubletap, focus, keypressed, longclick, longpress, pinch, postlayout,
-             singletap, swipe, touchcancel, touchend, touchmove, touchstart, twofingertap]
```
Did you validate the events work?

AngelkPetkov replied:
Yeah ,  all the ones i didn't take out work. dblcick and all the touchEvents. Added an new method to check events, same one as button class. As its just 2 buttons inside the view for the stepper.

hansemannn commented on apidoc/Titanium/UI/iOS/Stepper.yml:
```diff
@@ -48,28 +53,29 @@ properties:
+    summary: |
+        The miniumum value the stepper will be set to, the value must be smaller than the maximumValue.
+        If you attempt to set a value equal to or greater than maximumValue, the system will throw
+        an NSInvalidArgumentException exception.
```
`maximumValue`, also the `NSInvalidArgumentException` is not thrown and should be caught on the native side. Same with the property below.

hansemannn commented on iphone/Classes/TiUIiOSStepper.m:
```diff
@@ -28,107 +28,115 @@ -(UIStepper*)stepper
-    backgroundImageCache = [self imageWithImage:[self loadImage:value] fullScale:YES];
-    [[self stepper] setBackgroundImage:backgroundImageCache forState:UIControlStateNormal];
+    ENSURE_TYPE(value, NSNumber);
+    [stepper setEnabled:[TiUtils boolValue:value]];
```
Use `[self stepper]`

hansemannn commented on iphone/Classes/TiUIiOSStepper.m:
```diff
@@ -28,107 +28,115 @@ -(UIStepper*)stepper
 }
 
--(void)setBackgroundSelectedImage_:(id)value
+-(double)getValue
```
Objective-C getter don't have the `get`-Prefix, so just `- (double)value`. Why do we need this?

hansemannn commented on iphone/Classes/TiUIiOSStepper.m:
```diff
@@ -160,21 +168,54 @@ - (IBAction)stepperChanged:(id)sender
     if ([self.proxy _hasListeners:@"change"]) {
-        [self.proxy fireEvent:@"change" withObject:[NSDictionary dictionaryWithObject:newValue forKey:@"value"]];
+        [self.proxy fireEvent:@"change" withObject:[NSDictionary dictionaryWithObjectsAndKeys:
+                                                    [NSNumber numberWithDouble:stepper.value],@"value",
```
You can use `NUMDOUBLE()` here, makes it easier to read.

hansemannn commented on iphone/Classes/TiUIiOSStepper.m:
```diff
@@ -160,21 +168,54 @@ - (IBAction)stepperChanged:(id)sender
+            break;
+        default:
+            return;
+    }
```
Do we need to do this manually? Want to make sure we don't use business-logic multiple times

AngelkPetkov replied:
Yeah we do , other wise the events don't get called. Its the same used inside the button class and the stepper are just 2 buttons inside a view as i mentioned above.

hansemannn commented on iphone/Classes/TiUIiOSStepper.m:
```diff
@@ -160,21 +168,54 @@ - (IBAction)stepperChanged:(id)sender
+            return;
+    }
+    //[self setHighlighting:button.highlighted];
+    NSMutableDictionary *evt = [NSMutableDictionary dictionaryWithDictionary:[TiUtils touchPropertiesToDictionary:touch andPoint:[touch locationInView:self]]];
```
Just as a hint: I love to use `NSMutableDictionary *evt = [[TiUtils touchPropertiesToDictionary:touch andPoint:[touch locationInView:self]] mutableCopy];` here, but doesn't make any difference.

hansemannn commented on iphone/Classes/TiUIiOSStepper.m:
```diff
@@ -160,21 +168,54 @@ - (IBAction)stepperChanged:(id)sender
-                                                   dictionaryWithDictionary:[TiUtils touchPropertiesToDictionary:touch
-                                                                                andPoint:[touch locationInView:self]]]];
+    NSString *fireEvent;
+    NSString * fireActionEvent = nil;
```
Typo

hansemannn commented on iphone/Classes/TiUIiOSStepper.m:
```diff
@@ -160,21 +168,54 @@ - (IBAction)stepperChanged:(id)sender
+            touchStarted = NO;
+            fireEvent = @"touchend";
+            if (stepper.highlighted) {
+                fireActionEvent = [touch tapCount] == 1 ? @"click" : ([touch tapCount] == 2 ? @"dblclick" : nil);
```
Difference between `fireEvent` and `fireActionEvent` here? Both could be fired with `touchend` and `click`

AngelkPetkov replied:
We need need both of the string property for touchend to work along side click and dblclick. If only one property is used the fireActionEvent will get set to click and dblclick always as the button will be highlighted.

hansemannn commented on iphone/Classes/TiUIiOSStepperProxy.m:
```diff
@@ -35,15 +31,10 @@ -(TiDimension)defaultAutoHeightBehavior:(id)unused
 {
-    NSNumber *value = [NSNumber numberWithDouble:[[(TiUIiOSStepper*)[self view] stepper] value]];
-    return value;
+    return NUMDOUBLE([(TiUIiOSStepper*)[self view] getValue]);
```
Ok, here we use it. You could just return `NUMDOUBLE([[self stepper] value])`, or am I missing something?

AngelkPetkov replied:
The proxy doesnt have a refrence to the stepper you cant do [self stepper]

hansemannn commented:
- `stepValue` is not documented, also think of renaming to `steps`
- `value` setter is missing (currently creation only)
- `continuous` / `autorepeat` / `wraps` missing default values (check native docs)

AngelkPetkov commented:
Addressed Comments , i did not add a setValue method as it will be called every time you increment/decrement, we will need to check if the value passed in is greater than the min and larger than max. Which will cause an issue when incrementing because once you reach the max it will be printed. IF you're incrementing by 3 for example and the max is 29.

hansemannn commented on apidoc/Titanium/UI/iOS/Stepper.yml:
```diff
@@ -48,28 +53,35 @@ properties:
-        minimumValue sets value to maximumValue. If NO, the stepper does not increment beyond maximumValue nor
-        does it decrement below minimumValue but rather holds at those values.
+        If YES, incrementing beyond <Ti.UI.iOS.Stepper.maximum> sets value to <Ti.UI.iOS.Stepper.minimum>. likewise, decrementing below
+        <Ti.UI.iOS.Stepper.minimum> sets value to <Ti.UI.iOS.Stepper.maximum>. If NO, the stepper does not increment beyond <Ti.UI.iOS.Stepper.maximum> nor
```
Shouldnt it set the value to minimum of we set it below the minimum? Same with the maximum, make sure the bounds are correct.

AngelkPetkov replied:
This is correct, think of it like a cycle if you keep going up and go past the maximum value. You start from the beginning. Its not a validation where if ,for example the value is less than 0 we set it to 0.

hansemannn commented on apidoc/Titanium/UI/iOS/Stepper.yml:
```diff
@@ -48,28 +53,35 @@ properties:
-    summary: The miniumum value the stepper will be set to, the value must be smaller than the maximumValue.
+  - name: minimum
+    summary: |
+        The miniumum value the stepper will be set to, the value must be smaller than the maximumValue.
```
Typo in "miniumum"

hansemannn commented on apidoc/Titanium/UI/iOS/Stepper.yml:
```diff
@@ -48,28 +53,35 @@ properties:
-    summary: The maximum value the stepper will be set to, the value must be greater than the miniumumValue.
+  - name: maximum
+    summary: |
+        The maximum value the stepper will be set to, the value must be greater than the miniumumValue.
```
Typo in "miniumum", also "minimum", not "minimumVaue"

hansemannn commented on apidoc/Titanium/UI/iOS/Stepper.yml:
```diff
@@ -48,28 +53,35 @@ properties:
-    summary: The value the stepper will decrement and increment by.
+  - name: steps
+    summary: |
+      The value the stepper will increment and decrement by, default value for this property is 1.When setting
```
Typo, missing space.

hansemannn commented on apidoc/Titanium/UI/iOS/Stepper.yml:
```diff
@@ -108,21 +103,14 @@ properties:
-        Background image for the stepper decrement button in its dissabled state, specified as a local
-        file path or URL.
+        Background image for the stepper decrement button in its disabled state, specified as a local
+        file path or URL. The decrement button enteres a disabled state ones the value is equal to the
```
Typo, it's "enters"

hansemannn commented on apidoc/Titanium/UI/iOS/Stepper.yml:
```diff
@@ -132,39 +120,67 @@ properties:
 
 examples:
-  - title: Simple stepper Example
+  - title: Simple Stepper Example
```
Typo: "Simple stepper example"

hansemannn commented on apidoc/Titanium/UI/iOS/Stepper.yml:
```diff
@@ -132,39 +120,67 @@ properties:
+            <Alloy>
+                <Window class="container">
+                    <Stepper id="stepper" onClick="doClick" 
+                         top="30" tintColor="green" steps = "3"  />
```
Attributes in one line, no space in `steps = "3"`

hansemannn commented:
Left some comments. Regarding the `setValue`: It is important to be exposed because it's **very** common to set the value manually after creating the control. We need to take care of that and do the bounds check accordinately.

AngelkPetkov commented:
Addressed Comments , also added the setValue. Disregard what i said before, forgot to turn notification to NO, which fixes the issue of setValue being called upon increment/decrement.

AngelkPetkov committed:
Merge branch 'master' of https://github.com/appcelerator/titanium_mobile into TIMOB-6305

hansemannn commented on apidoc/Titanium/UI/iOS/Stepper.yml:
```diff
@@ -0,0 +1,186 @@
+
+    A Stepper displays its selected state while it is being pressed.
+    You can specify background images for each state on iOS.
+    Use the <Titanium.UI.createStepper> method or a **`<stepper>`** Alloy element to create a stepper
```
It's `<stepper>` (capitalized), Alloy wouldn't be able to handle it if not.

AngelkPetkov committed:
[TIMOB-6305] iOS: Addressed Comment

hansemannn commented:
Approved!
